s3/costs/prepare_response.go

```diff
@@ -0,0 +1,149 @@
+	bucketInfo.GbMonth = bucketData["usage"].(bucket)["value"].(float64)
+	bucketInfo.StorageCost = bucketData["cost"].(bucket)["value"].(float64)
+	return bucketInfo
+}
+
+// getBucketRequests gets the requests informations from the bucketData
+// and returns an *S3BucketCost filled with these data
+func getBucketRequests(bucketInfo *S3BucketCost, bucketData bucket) *S3BucketCost {
+	bucketInfo.Requests = bucketData["usage"].(bucket)["value"].(float64)
+	bucketInfo.RequestsCost = bucketData["cost"].(bucket)["value"].(float64)
+	return bucketInfo
+}
+
+// getBucketBandwidthIn gets the bandwidthIn informations from the bucketData
+// and returns an *S3BucketCost filled with these data
+func getBucketBandwidthIn(bucketInfo *S3BucketCost, bucketData bucket) *S3BucketCost {
+	bucketInfo.DataOut = bucketData["usage"].(bucket)["value"].(float64)
+	bucketInfo.BandwidthCost += bucketData["cost"].(bucket)["value"].(float64)
+	return bucketInfo
+}
+
+// getBucketBandwidthOut gets the bandwidthOut informations from the bucketData
+// and returns an *S3BucketCost filled with these data
+func getBucketBandwidthOut(bucketInfo *S3BucketCost, bucketData bucket) *S3BucketCost {
+	bucketInfo.DataIn = bucketData["usage"].(bucket)["value"].(float64)
+	bucketInfo.BandwidthCost += bucketData["cost"].(bucket)["value"].(float64)
+	return bucketInfo
+}
+
+// parseBuckets iterates through all the buckets and calls the getter function corresponding
+// to the resultType
+func parseBuckets(buckets bucketsInfo, parsedDocument bucket, resultType string) bucketsInfo {
+	bucketsField := parsedDocument["buckets"].([]interface{})
+	for _, bucketData := range bucketsField {
+		bucketData := bucketData.(bucket)
+		bucketName := bucketData["key"].(string)
+		// The billing data can contain billings for errored requests that we do not want to see
+		if isValidBucket(bucketName) {
+			bucketInfo := getBucketInfoByName(buckets, bucketName)
+			if resultTypePtr, ok := resultTypeToBucketCostGetter[resultType]; ok {
+				bucketInfo = resultTypePtr(bucketInfo, bucketData)
+			}
+		}
+	}
+	return buckets
+}
+
+// parseESResult parses an *elastic.SearchResult according to it's resultType
+func parseESResult(ctx context.Context, buckets bucketsInfo, res *elastic.SearchResult, resultType string) (bucketsInfo, error) {
+	var logger = jsonlog.LoggerFromContextOrDefault(ctx)
+	var parsedDocument bucket
+	err := json.Unmarshal(*res.Aggregations["buckets"], &parsedDocument)
+	if err != nil {
+		logger.Error("Failed to parse elasticsearch document.", err.Error())
+		return buckets, err
+	}
+	buckets = parseBuckets(buckets, parsedDocument, resultType)
+	return buckets, nil
+}
+
+// prepareResponse parses the results from elasticsearch and returns a map of buckets with their usage informations
+func prepareResponse(ctx context.Context, resStorage, resRequests, resBandwidthIn, resBandwidthOut *elastic.SearchResult) (interface{}, error) {
```

Comment:
Don’t repeat yourself! What about something like this?
```go
func prepareResponse(
    ctx context.Context,
    resStorage, resRequests, resBandwidthIn, resBandwidthOut *elastic.SearchResult
) (interface{}, error) {
    var buckets = make(bucketsInfo)
    var err error
    var components = [...]struct{k string, sr *elastic.SearchResult}{
        { "storage", resStorage },
        { "requests", resRequests },
        { "bandwidthIn", resBandwidthIn },
        { "bandwidthOut", resBandwidthOut },
    }
    for _, cpn := range components {
        buckets, err = parseESResult(ctx, buckets, cpn.sr, cpn.k)
        if err != nil {
           return nil, err
        }
    }
    return buckets, nil
}
```

Reply:
Done!